### PR TITLE
MSVC C++20 Support

### PR DESCRIPTION
**Pull request by sean-parent**
Added Support for MSVC C++20.
Fixed warnings at \W4
Improved cmake presets support for VSCode.
New libraries:
- macro_utilities.hpp
  - ADOBE_STRINGIZE(x)
  - ADOBE_LINE_STRING()
  - ADOBE_MESSAGE(message)
- exception.hpp
  - terminate(message)
- append.hpp
  - append(container, range)
- unique.hpp
  - sort_unique(container)

### CODE REVIEW

File: test/adam_smoke/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,13 @@
 add_executable (adam_smoke_test adam_smoke_test.cpp)
 target_link_libraries(adam_smoke_test PRIVATE asl)
 
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
+    target_compile_options (adam_smoke_test PRIVATE "/DNOMINMAX")
+    target_compile_options (adam_smoke_test PRIVATE "/permissive-")
+    target_compile_options (adam_smoke_test PRIVATE "/Zc:__cplusplus")
+    target_compile_options (adam_smoke_test PRIVATE "/W4")
+    target_compile_options (adam_smoke_test PRIVATE "/WX")
+endif()
```
Comment (sean-parent) [inline test/adam_smoke/CMakeLists.txt:10]: @dabrahams @camio - There must be a better way to manage compiler settings.
I want the library to require the first 3 options, and the cilents (test cases) to add the second 2 options.

File: adobe/manip.hpp [hunk 2]
```diff
@@ -18,11 +18,17 @@
 
 #include <iostream>
 
+/*
+    REVISIT (sean-parent): As far as I can determine, this file does not use adobe/serializable.hpp
+    However, if it is not included here, the build fails for C++17 under MSVC only. I have not been
+    able to determine why this is the case.
+*/
+#include <adobe/serializable.hpp>
```
Comment (sean-parent) [inline adobe/manip.hpp:26]: @dabrahams - I was able to significantly reduce the problem to the pointer where I determined that `adobe/serializable.hpp` had to be included before `manip.hpp`, but there does not appear to be any dependency on `serializable.hpp` - circular or not. In fact, in my reduced case, not include `manip.hpp` at all also worked. If it was more than one configuration of one compiler, I'd reduce it more.

Wihtout this line the following fails to compiler:
```
#include <adobe/array.hpp>
#include <adobe/manip.hpp>
#include <adobe/serializable.hpp>

/**************************************************************************************************/

namespace adobe {

/**************************************************************************************************/

void test() {
    serializable_t value{array_t{}};
}

/**************************************************************************************************/

} // namespace adobe
```

File: test/adam_smoke/CMakeLists.txt (hunk 1 above)
Comment (dabrahams) [reply test/adam_smoke/CMakeLists.txt:10]: If by
> I want the library to require the first 3 options  and the cilents (test cases) to add the second 2 options.

You actually mean you want the library to be built with the first 3 options and its clients to have the 2nd 2 options imposed on them automatically due to the fact that they link the library, then you want

```cmake
target_compile_options(the_library PRIVATE  "/Zc:__cplusplus" PUBLIC "/W4" "/WX")
```

The last two become usage requirements as well as requirements for building the library. 

@camio's statement that:

> When the library writer dictates flags, the library gets tied to a snapshot of today's compilers and options making it unportable

Could be interpreted two ways: 1, you should try to use higher-level constructs than the flags for specific compilers whenever possible, or 2, you should never employ usage requirements.  I agree with the first but would disagree with the second.  I'm not sure what these flags do or whether higher-level constructs exist for them.

File: adobe/manip.hpp (hunk 2 above)
Comment (dabrahams) [reply adobe/manip.hpp:26]: Without what line?

I don't think giving up on the reduction is appropriate.  This configuration of this compiler applies all Adobe products and many others.  IMO you should get to the point where you can submit a bug report.  But that's just me.

File: .vscode/settings.json [hunk 3]
```diff
@@ -1,136 +1,118 @@
...
-        "*.ipp": "cpp",
-        "*.tcc": "cpp",
-        "rope": "cpp",
-        "slist": "cpp",
-        "print": "cpp"
-    },
-    "C_Cpp.default.compilerPath": "/usr/bin/clang++"
-}
+{
+    "files.associations": {
```
Comment (dabrahams) [inline .vscode/settings.json:2]: I don't know how to analyze this change.  Surely it's not part of ”MSVC C++20 Support“ though.

File: CMakeLists.txt [hunk 4]
```diff
@@ -7,8 +7,7 @@ project(adobe_source_libraries CXX)
 
 set(CMAKE_CXX_EXTENSIONS OFF)
 if(NOT DEFINED CMAKE_CXX_STANDARD)
-    set(CMAKE_CXX_STANDARD 17)
-    set(CMAKE_XCODE_ATTRIBUTE_CLANG_CXX_LANGUAGE_STANDARD "c++17")
+    set(CMAKE_CXX_STANDARD 20)
```
Comment (dabrahams) [inline CMakeLists.txt:10]: I'm going to assume that dropping the Xcode setting is intentional and correct.

File: CMakePresets.json [hunk 5]
```diff
@@ -3,23 +3,51 @@
...
                 "CMAKE_BUILD_TYPE": "Debug",
+                "CMAKE_CXX_STANDARD": "20",
+                "CMAKE_INSTALL_PREFIX": "${sourceDir}/out/install/${presetName}",
+                "VCPKG_TARGET_TRIPLET": "x64-windows"
+            },
+            "architecture": {
+                "value": "x64",
+                "strategy": "external"
+            },
+            "toolchainFile": "$env{VCPKG_ROOT}/scripts/buildsystems/vcpkg.cmake",
```
Comment (dabrahams) [inline CMakePresets.json:20]: So this toolchain requires vcpkg?  Is that dependency documented in the README?  If not you better do that.
Is how to use the presets documented?

File: adobe/exception.hpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+#include <string>
+
+/**************************************************************************************************/
+
+namespace adobe {
+
+/**************************************************************************************************/
+
+/// invoke `std::terminate` with the given message.
```
Comment (dabrahams) [inline adobe/exception.hpp:19]: ? You can't invoke `std::terminate` with a message, so you need to fix this.

File: adobe/implementation/toroid.hpp [hunk 7]
```diff
@@ -227,7 +227,7 @@ class byte_toroid_t : boost::noncopyable {
 
     void purify(toroid_node_t* p) {
         toroid_header_t* c(column_of(p));
-        int x(p->color_m);
+        auto x(p->color_m);
```
Comment (dabrahams) [inline adobe/implementation/toroid.hpp:230]: Surely not needed for C++20 support?  Why is it even an improvement?  

Aside: we really don't use copy-initialization anymore?  This syntax makes it look like a function declaration.

File: adobe/iomanip.hpp [hunk 8]
```diff
@@ -16,6 +16,8 @@
 
 #include <adobe/config.hpp>
 
+#include <adobe/iomanip_fwd.hpp>
```
Comment (dabrahams) [inline adobe/iomanip.hpp:19]: If header order matters, you'd better leave a big-@ss comment telling people not to move the header.  Should really refer to a bug report and say that it's a workaround.

File: adobe/json.hpp [hunk 9]
```diff
@@ -200,11 +200,10 @@ class json_parser {
         frac();
         exp();
 
-        const auto p_len = p_ - p;
         char* p_end = nullptr;
         double value = std::strtod(p, &p_end);
         require(std::isfinite(value), "finite number");
-        ADOBE_ASSERT(p_len == (p_end - p) && "std::strtod() failure");
+        ADOBE_ASSERT((p_ - p) == (p_end - p) && "std::strtod() failure");
```
Comment (dabrahams) [inline adobe/json.hpp:206]: Surely not needed for C++20?  Is this an improvement?  It's not great code either way, but at least the old code let you know that `p_ - p` is somehow the length of `p` (whatever that is).

File: adobe/macro_utilities.hpp [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+#define ADOBE_STRINGIZE(x)                                                                         \
+    ADOBE_DETAIL_STRINGIZE2(x)        // NOLINTNEXTLINE(cppcoreguidelines-macro-usage)
+#define ADOBE_DETAIL_STRINGIZE2(x) #x // NOLINT(cppcoreguidelines-macro-usage)
+
+
+/// a string literal for __LINE__.
+#define ADOBE_LINE_STRING() ADOBE_STRINGIZE(__LINE__) // NOLINT(cppcoreguidelines-macro-usage)
+
+/// Constructs a [GNU error formatted](https://www.gnu.org/prep/standards/standards.html#Errors)
+/// message string. Message must be a string literal.
```
Comment (dabrahams) [inline adobe/macro_utilities.hpp:21]:
```suggestion
/// Expands to a [GNU error formatted](https://www.gnu.org/prep/standards/standards.html#Errors)
/// string literal reporting `message` at the line where it is invoked.
///
/// - Requires: `message` is a string literal.
```

File: adobe/manip.hpp [hunk 11]
```diff
@@ -18,11 +18,17 @@
 
 #include <iostream>
 
+/*
+    REVISIT (sean-parent): As far as I can determine, this file does not use adobe/serializable.hpp
+    However, if it is not included here, the build fails for C++17 under MSVC only. I have not been
```
Comment (dabrahams) [inline adobe/manip.hpp:23]: If include order matters in this file, I think you could be clearer about "here."  Could be interpreted to mean “in this file.”

File: adobe/serializable.hpp [hunk 12]
```diff
@@ -14,13 +14,16 @@
 #include <ostream>
 #include <type_traits>
 
-#include <adobe/typeinfo.hpp>
 #include <adobe/string/to_string.hpp>
+#include <adobe/typeinfo.hpp>
```
Comment (dabrahams) [inline adobe/serializable.hpp:18]: Is something automatically reordering these?  How does that play with your include order constraints?

File: adobe/serializable.hpp [hunk 13]
```diff
@@ -63,10 +66,10 @@ template <class T>
 inline typename std::enable_if<!has_ostream_insertion<T>::value>::type
 ostream_insertion(std::ostream&, const T&) {}
 
-#else // __cplusplus < 202002L
+#else  // __cplusplus < 202002L
 
 template <class T>
-void ostream_insertion(std::ostream& s, const T& x) {
+inline void ostream_insertion(std::ostream& s, const T& x) {
```
Comment (dabrahams) [inline adobe/serializable.hpp:72]: Why this change?

File: adobe/serializable.hpp [hunk 14]
```diff
@@ -110,7 +113,7 @@ class serializable_t {
 
     serializable_t(serializable_t&& x) : instance_m(x.instance_m.release()) {}
 
-    void operator()(std::ostream& s) const { object()(s); }
+    void operator()(std::ostream& s) const { object()._out(s); }
```
Comment (dabrahams) [inline adobe/serializable.hpp:116]: If you're gonna change stuff you should give it a doc comment along the way; just sayin'.

File: adobe/sha.hpp [hunk 15]
```diff
@@ -239,8 +241,9 @@ std::uint64_t stuff_into_state(typename HashTraits::message_block_type& state,
             v <<= message_end - 8;
 
         *dst |= v;
-        stuff_bit_offset += num_bits;
-        n_stuffed_bits += num_bits;
+        ADOBE_ASSERT(num_bits < 8);
+        stuff_bit_offset += static_cast<std::uint16_t>(num_bits);
```
Comment (dabrahams) [inline adobe/sha.hpp:245]: Maybe a comment explaining how you know it won't overflow is warranted.

File: source/adam.cpp [hunk 16]
```diff
@@ -1114,36 +1114,34 @@ void sheet_t::implementation_t::flow(cell_bits_t& priority_accessed) {
                 index_t::iterator iter = output_index_m.find(term->name_set_m[n]);
                 assert(iter != output_index_m.end());
 
-                cell_t& cell = *iter;
+                cell_t& e = *iter;
```
Comment (dabrahams) [inline source/adam.cpp:1117]: OMG this is a hecka long function.  Nice getting rid of the type information in the name, but can we do better than `e`?  What role is this variable playing on the inside of the (raw!) loop?  That is, if you were to break the body of the loop into a function call taking `cell_t&`, what would you call the parameter?

File: source/eve.cpp [hunk 17]
```diff
@@ -1347,11 +1347,11 @@ void view_proxy_t::solve_up_with(::child_iterator first, ::child_iterator last,
         accumulate function object.
         */
 
-        for (guide_set_t::iterator first(boost::next(forward_guide_set.begin())),
-             last(forward_guide_set.end());
-             first != last; ++first) {
+        for (guide_set_t::iterator f(boost::next(forward_guide_set.begin())),
```
Comment (dabrahams) [inline source/eve.cpp:1350]: Is this an improvement?

File: source/eve_evaluate.cpp [hunk 18]
```diff
@@ -325,9 +326,9 @@ void apply_layout_parameters(layout_attributes_t& data, const dictionary_t& para
 
                 layout_attributes_t::spacing_t::iterator dest_iter(data.spacing_m.begin() + 1);
 
-                for (array_t::const_iterator iter(spacing_array.begin());
-                     iter != spacing_array.end(); ++iter) {
-                    *dest_iter = iter->cast<int>();
+                for (array_t::const_iterator spacing_iter(spacing_array.begin());
```
Comment (dabrahams) [inline source/eve_evaluate.cpp:329]: I think a single character name would actually be clearer in this case.  Also, though, no raw loops would alleviate the need to name these trivial things everywhere.

File: test/property_model_eval/iomanip_flat.cpp [hunk 19]
```diff
@@ -151,7 +151,7 @@ void flat_format::handle_atom(stream_type& os, bool is_push) {
         } else if (value.type_info() == typeid(dictionary_t)) {
             os << value.cast<dictionary_t>();
         } else if (value.type_info() == typeid(array_t)) {
-            os << value.cast<array_t>();
+            os << top.value().cast<::adobe::array_t>();
```
Comment (dabrahams) [inline test/property_model_eval/iomanip_flat.cpp:154]: Probably you want to revert this change, but if not, comment to explain why this line needs to be so different from the rest.

**Review by dabrahams — CHANGES_REQUESTED** [review]
Some suggestions inline.

File: adobe/algorithm/append.hpp [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace adobe {
+
+/**************************************************************************************************/
+
+/// Insert the range `r` at the end of the sequence container `x`.
+/// Returns an iterator pointing to the first element inserted, or `x.end()` if `r` is empty.
+template <class T, // T models SequenceContainer
+          class R> // R models Range
+inline auto append(T& x, const R& r) {
```
Comment (dabrahams) [inline adobe/algorithm/append.hpp:21]:
```suggestion
```suggestion
/// Inserts the elements of *range* `r` (a `Range`) at the end of *forward container* `x`,
/// returning the position corresponding to the incoming `end` of `x`.
template <class T, class R>
inline auto append(T& x, const R& r) {
```

File: source/adam.cpp [hunk 21]
```diff
@@ -345,6 +346,27 @@ class sheet_t::implementation_t : boost::noncopyable {
     priority_t name_to_priority(name_t name) const;
     void flow(cell_bits_t& priority_accessed);
 
+    /// Returns the output cell with the given name.
+    /// \pre the output cell must exist.
```
Comment (dabrahams) [inline source/adam.cpp:350]:
```suggestion
    /// Returns the output cell with the given name.
    ///
    /// \pre A cell with that name exists.
```

Blank line after summary.
A precondition is a predicate statement.  Write in English what you'd put in the `assert()` parens.

File: source/adam.cpp [hunk 22]
```diff
@@ -345,6 +346,27 @@ class sheet_t::implementation_t : boost::noncopyable {
...
+        assert(p != output_index_m.end() && "output cell not found");
+        return *p;
+    }
+    const cell_t& output_cell(const name_t& name) const {
+        auto p = output_index_m.find(name);
+        assert(p != output_index_m.end() && "output cell not found");
+        return *p;
+    }
+
+    /// Returns `true` if any output cells in the relation are resolved, false otherwise.
```
Comment (dabrahams) [inline source/adam.cpp:362]: Some people think iff is too obscure, but this is an option

```suggestion
    /// Returns `true` iff any output cells in the relation are resolved.
```
Some people like:
```suggestion
    /// Returns whether any output cells in the relation are resolved.
```

I find that really awkward personally but I can see the argument.

If you are in the context of a type representing a relation, you can leave out “in the relation.”

File: test/adam_smoke/CMakeLists.txt (hunk 1 above)
Comment (sean-parent) [reply test/adam_smoke/CMakeLists.txt:7]: I updated it to use PRIVATE/PUBLIC (more) appropriately.

File: adobe/implementation/toroid.hpp (hunk 7 above)
Comment (sean-parent) [reply adobe/implementation/toroid.hpp:230]: This silences a warning because `x` is char type and later uses will narrow back to a char type.

File: adobe/sha.hpp (hunk 15 above)
Comment (sean-parent) [reply adobe/sha.hpp:245]: I'm not an expert in this code - it took me a bit to figure out that `num_bits` < 8 here. I'm not certain, but it looks like `stuff_bit_offset` is always used in modulo value_bitsize_k and can warp (unsigned). `n_stuffed_bits` is 64 bits, so it would take a lot to overflow incrementing by `8`. And I _think_ this is going to be bounded by sizeof message_block_type * 8 - so you'll run out of VM space first (if not addressable space).

Comment (dabrahams) [reply adobe/sha.hpp:245]: Well, if you can't prove it, there's always `boost::numeric_cast` to maintain safety.

File: source/adam.cpp (hunk 16 above)
Comment (sean-parent) [reply source/adam.cpp]: I did a fair amount of refactoring and cleanup of that function and used better names. Still long - opened Issue#127 to revisit.

File: adobe/sha.hpp (hunk 15 above)
Comment (sean-parent) [reply adobe/sha.hpp:245]: That might be very bad if it is expected to wrap.

File: source/adam.cpp (hunk 21 above)
Comment (sean-parent) [reply source/adam.cpp:350]: Accepted.

File: source/adam.cpp (hunk 22 above)
Comment (sean-parent) [reply source/adam.cpp:362]: My preference was iff (I had that initially, but I know people who would object, so I changed it).  I'll go with "whether" -

File: adobe/algorithm/append.hpp (hunk 20 above)
Comment (sean-parent) [reply adobe/algorithm/append.hpp]: Accepted.

File: adobe/manip.hpp (hunk 2 above)
Comment (sean-parent) [reply adobe/manip.hpp]: Found it - a namespace lookup issue. Fixed property in array.hpp.

File: .vscode/settings.json (hunk 3 above)
Comment (sean-parent) [reply .vscode/settings.json:2]: This file really shouldn't be here - I'll do the two step delete and add to .git-ignore.

File: adobe/exception.hpp (hunk 6 above)
Comment (sean-parent) [reply adobe/exception.hpp]: I updated the comment to be more clear.

File: CMakeLists.txt (hunk 4 above)
Comment (sean-parent) [reply CMakeLists.txt:10]: Yes. It isn't needed.

File: CMakePresets.json (hunk 5 above)
Comment (sean-parent) [reply CMakePresets.json:20]: The system is currently using vcpkg to pull in boost. It's undocumented, and I'd like to get rid of it in favor of fetch content, but not in this PR.

File: source/eve_evaluate.cpp (hunk 18 above)
Comment (sean-parent) [reply source/eve_evaluate.cpp]: Replaced this one with `transform()`

File: source/eve.cpp (hunk 17 above)
Comment (sean-parent) [reply source/eve.cpp:1350]: Was a quick way to squelch a warning - but I resolved the revisit comment and moved the code to use std::generate()

File: adobe/json.hpp (hunk 9 above)
Comment (sean-parent) [reply adobe/json.hpp:206]: The unused `p_len` was generating warnings in release builds.

File: adobe/serializable.hpp (hunk 12 above)
Comment (sean-parent) [reply adobe/serializable.hpp:18]: No, include order constraints - and yes, clang-format alphabetizes the groups.

File: adobe/serializable.hpp (hunk 13 above)
Comment (sean-parent) [reply adobe/serializable.hpp:72]: "inline"? Because it is an empty or single-line function in a header. It doesn't mean much for a template.

File: adobe/serializable.hpp (hunk 14 above)
Comment (sean-parent) [reply adobe/serializable.hpp:116]: Good point - I'll try and do better.
